File: android/app/src/test/java/com/example/yugiohscanner/PriceStepsTest.kt
package com.example.yugiohscanner

import com.example.yugiohscanner.ml.PriceRef
import com.example.yugiohscanner.ml.PriceSteps
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Test

/** ZWILLING von desktop/src/utils/priceSteps.test.js -- dieselbe Fixture docs/fixtures/portfolio/price-steps.json. */
class PriceStepsTest {
    @Test fun `alle Fixture-Faelle`() {
        val cases = JSONObject(Fixtures.text("docs/fixtures/portfolio/price-steps.json")).getJSONArray("cases")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val name = c.getString("name")
            val rowsJson = c.getJSONArray("rows")
            val rows = (0 until rowsJson.length()).map { j ->
                val o = rowsJson.getJSONObject(j)
                PriceRef("1", "X", "DE", "Common", o.getString("day"), o.getDouble("price"), o.getString("source"))
            }
            val s = PriceSteps.compute(rows, c.getString("today"), c.getInt("window"))
            val e = c.getJSONObject("expected")
            assertEquals(name, e.getString("kind"), s.kind)
            if (e.has("flatDay")) {
                assertEquals(name, e.getString("flatDay"), s.flatDay)
                assertEquals(name, e.getDouble("flatPrice"), s.flatPrice!!, 1e-9)
            }
            val ep = e.getJSONArray("points")
            assertEquals("$name points", ep.length(), s.points.size)
            s.points.forEachIndexed { j, p ->
                assertEquals("$name points[$j]", ep.getJSONObject(j).getString("day"), p.day)
                assertEquals("$name points[$j]", ep.getJSONObject(j).getDouble("price"), p.price, 1e-9)
            }
            val em = e.getJSONArray("markers")
            assertEquals("$name markers", em.length(), s.markers.size)
            s.markers.forEachIndexed { j, m ->
                assertEquals("$name markers[$j]", em.getJSONObject(j).getString("day"), m.day)
                assertEquals("$name markers[$j]", em.getJSONObject(j).getString("family"), m.family)
            }
        }
    }
}

File: desktop/src/utils/priceSteps.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { computeSteps, familyOfSource, addDaysUtc, fmtDayDE, todayUtc } from './priceSteps.js';

// ZWILLING: android/app/src/test/java/com/example/yugiohscanner/PriceStepsTest.kt liest dieselbe Fixture.
const FIX = JSON.parse(readFileSync(new URL('../../../docs/fixtures/portfolio/price-steps.json', import.meta.url), 'utf8'));

for (const c of FIX.cases) {
  test(`Fixture: ${c.name}`, () => {
    assert.deepEqual(computeSteps(c.rows, c.today, c.window), c.expected);
  });
}

test('Hilfen: Familie, Tage, Kurzdatum, heute', () => {
  assert.equal(familyOfSource('cm_scrape'), 'cm');
  assert.equal(familyOfSource(undefined), 'unknown');
  assert.equal(addDaysUtc('2026-03-01', -1), '2026-02-28');
  assert.equal(fmtDayDE('2026-09-23'), '23.09.');
  assert.equal(todayUtc(new Date('2026-09-20T23:30:00-02:00')), '2026-09-21');
});
